src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -0,0 +1,353 @@
+    this.numberOfSkippedLines = MoreObjects.firstNonNull(numberOfSkippedLines, 0);
+
+    this.partitionStartByteRange = partitionStartByteRange;
+    this.partitionEndByteRange = partitionEndByteRange;
+    this.isLastWorker = isLastWorker;
+  }
+
+  @Override
+  protected TupleBatch fetchNextReady() throws IOException, DbException {
+    long lineNumberBegin = lineNumber;
+    boolean fixingStartByte = false;
+    boolean onLastRow = false;
+
+    boolean discardedRecord = false;
+
+    while ((buffer.numTuples() < TupleBatch.BATCH_SIZE)) {
+      discardedRecord = false;
+      lineNumber++;
+      if (parser.isClosed()) {
+        break;
+      }
+      try {
+        if (!iterator.hasNext()) {
+          parser.close();
+          break;
+        }
+      } catch (final RuntimeException e) {
+        throw new DbException("Error parsing row " + lineNumber, e);
+      }
+      CSVRecord record = iterator.next();
+      // This covers the case where the first row of a worker matches the schema. We only want to read this row if the
+      // previous character is '\n' or '\r'
+      if (record.size() == schema.numColumns()
+          && lineNumber - 1 == 0
+          && partitionStartByteRange != 0) {
+        InputStreamReader startStreamReader =
+            new InputStreamReader(
+                source.getInputStream(partitionStartByteRange - 1, partitionEndByteRange));
+        char currentChar = (char) startStreamReader.read();
+        if (currentChar != '\n' && currentChar != '\r') {
+          discardedRecord = true;
+        }
+      }
+      // This covers the case where the last row matches the schema's number of columns. We need to ensure that we read
+      // the entire row completely
+      else if (record.size() == schema.numColumns() && !iterator.hasNext() && !isLastWorker) {
+        long movingEndByte = partitionEndByteRange;
+        long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
+        boolean newLineFound = false;
+        while (!newLineFound) {
+          movingEndByte += MyriaConstants.BYTE_OVERLAP_PARALLEL_INGEST;
+          // Create a stream to look for the new line
+          InputStream trailingEndInputStream =
+              source.getInputStream(partitionEndByteRange, movingEndByte);
+          InputStreamReader startStreamReader = new InputStreamReader(trailingEndInputStream);
+          int dataChar = startStreamReader.read();
+          while (dataChar != -1) {
+            char currentChar = (char) dataChar;
+            if (currentChar == '\n' || currentChar == '\r') {
+              newLineFound = true;
+              // Re-initialize the parser with the last row only
+              InputStream beginningOfRecord =
+                  source.getInputStream(bytePositionAtBeginningOfRecord, partitionEndByteRange);
+              InputStream concatenateEndOfRecord =
+                  source.getInputStream(partitionEndByteRange + 1, movingEndByte);
```

Comment:
This one is the one I'm stumped on and perhaps I'm misunderstanding something. I don't understand how we can simply reuse ```trailingEndInputStream```. We are reading through this stream to find a new line character. If we reuse this trailing stream for the sequence stream, how would that work? Our trailing stream would not start at ```partitionEndByteRange``` anymore since we've already read through it. 

Currently, I have a cleaner version of this code and I've fixed most of the other issues you've brought up. This is the only thing I am still struggling to understand. 

Thanks!